File: src/2580_统计将重叠区间合并成组的方案数.cpp
#include <vector>
#include <algorithm> // Include this for std::sort

class Solution
{
public:
    int countWays(std::vector<std::vector<int>> &ranges)
    {
        // Sort the vector of vectors based on the start points of the ranges
        std::sort(ranges.begin(), ranges.end(), [](const std::vector<int> &a, const std::vector<int> &b)
                  { return a[0] < b[0]; });

        const int MOD = 1'000'000'007;
        int ans = 1;
        int max_r = -1;
        for (auto &p : ranges)
        {
            if (p[0] > max_r)
            {
                // Found a non-overlapping interval, multiply the answer by 2
                ans = static_cast<long long>(ans) * 2 % MOD; // Use long long to avoid overflow
            }
            // Extend the current range to include the new interval
            max_r = std::max(max_r, p[1]);
        }
        return ans;
    }
};

File: src/2580_统计将重叠区间合并成组的方案数.rs
// 2580_统计将重叠区间合并成组的方案数
// https://leetcode.cn/problems/count-ways-to-group-overlapping-ranges/description/

impl Solution {
    pub fn count_ways(mut ranges: Vec<Vec<i32>>) -> i32 {
        ranges.sort_unstable();
        let n = ranges.len();
        let mut i = 0;
        let mut ans = 1;
        while i < n {
            let mut r = ranges[i][1];
            let mut j = i + 1;
            while j < n && ranges[j][0] <= r {
                if r < ranges[j][1] {
                    r = ranges[j][1];
                }
                j += 1;
            }
            ans = (ans as i64 * 2 % 1_000_000_007) as i32;
            i = j;
        }
        ans
    }
}
